Implement HttpContentAsyncStream.Write in CurlHandler

stephentoub opened the pull request:
WCF ends up trying to use Stream.Write as part of HttpContent.CopyToAsync implementation.  While that should be addressed separately to use WriteAsync, it's possible others are also using Write instead of WriteAsync, and today we throw NotSupportedException from Write.  So as to let this at least work, I'm implementing Write to delegate to WriteAsync. I also noticed we were missing a FlushAsync implementation, so I've added one.

cc: @davidsh, @roncain, @kapilash

davidsh commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.HttpContentAsyncStream.cs:
```diff
@@ -142,6 +142,15 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                // WriteAsync should be how writes are performed on this stream as
+                // part of HttpContent.CopyToAsync.  However, we implement Write just
```
Actually, derived classes of the abstract `HttpContent` class do not implement `CopyToAsync`.  That is a public method which actually calls protected methods such as `SerializeToStreamAsync`. That is the method that WCF and others override. So, I suggest that the comments be revised.

stephentoub replied:
Thanks, but I'd like to leave this as is.  The comment doesn't state anything about overriding CopyToAsync, but rather states that Write happens as part of CopyToAsync, which includes anything it invokes, and what it invokes is an implementation detail from the perspective of CurlHandler.  Since CurlHandler is simply invoking CopyToAsync, the comment is calling out that these writes happen as part of that.

davidsh commented:
Left some feedback about the comments. Otherwise LGTM.

stephentoub commented:
Based on a conversation with @mconnew, I've added an additional commit to help make sure that the Stream isn't used incorrectly, even though we're handing it out to an arbitrary HttpContent implementation.  Previously CanRead would return true, but since the Stream is only meant for writing external to CurlHandler's implementation, I've changed it to return false, and we now throw from the public Read/ReadAsync methods; internal usage uses a separate method for reading.

mconnew commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.HttpContentAsyncStream.cs:
```diff
@@ -142,6 +149,15 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
+                // part of HttpContent.CopyToAsync.  However, we implement Write just
+                // in case someone does need to do a synchronous write or has existing
+                // code that does so.
+                WriteAsync(buffer, offset, count, CancellationToken.None).GetAwaiter().GetResult();
```
Could this deadlock if there's a single threaded SynchronizationContext?  There's a nice article about it [here](http://blogs.msdn.com/b/pfxteam/archive/2011/01/13/10115163.aspx). :stuck_out_tongue:

stephentoub replied:
> There's a nice article about it here. :stuck_out_tongue:

Thanks.  Not sure the author really knows what he's talking about, though.

> Could this deadlock if there's a single threaded SynchronizationContext?

Nothing about the implementation should cause that kind of deadlock.  The implementation never schedules anything back to the original SynchronizationContext.

It could, however, block for an arbitrarily long period (it's pushing data which won't be consumed and allow the Write to wake up until libcurl asks for the data), which is why I didn't implement this initially, and why WCF shouldn't be using it :smile:

vijaykota commented:
LGTM
